website/source/guides/writing-custom-terraform-providers.html.md

```diff
@@ -0,0 +1,553 @@
+
+Terraform supports a plugin model, and all providers and actually plugins.
+Plugins are distributed as Go binaries. Although technically possible to write a
+plugin in another language, almost all Terraform plugins are written in
+[Go](https://golang.org). For more information on installing and configuring Go,
+please visit the [Golang installation guide](https://golang.org/doc/install).
+
+This post assumes familiarity with Golang and basic programming concepts.
+
+As a reminder, all of Terraform's core providers are open source. When stuck or
+looking for examples, please feel free to reference
+[the open source providers](https://github.com/hashicorp/terraform/tree/master/builtin/providers) for help.
+
+## The Provider Schema
+
+To start, create a file named `provider.go`. This is the root of the provider
+and should include the following boilerplate code:
+
+```go
+package main
+
+import (
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func Provider() *schema.Provider {
+	return &schema.Provider{
+		ResourcesMap: map[string]*schema.Resource{},
+	}
+}
+```
+
+The
+[`helper/schema`](https://godoc.org/github.com/hashicorp/terraform/helper/schema)
+library is part of Terraform's core. It abstracts many of the complexities and
+ensures consistency between providers. The example above defines an empty provider (there are no _resources_).
+
+The `*schema.Provider` type describes the provider's properties including:
+
+- the configuration keys it accepts
+- the resources it supports
+- any callbacks to configure
+
+## Building the Plugin
+
+Go requires a `main.go` file, which is the default executable when the binary is
+built. Since Terraform plugins are distributed as Go binaries, it is important
+to define this entry-point with the following code:
+
+```go
+package main
+
+import (
+	"github.com/hashicorp/terraform/plugin"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func main() {
+	plugin.Serve(&plugin.ServeOpts{
+		ProviderFunc: func() terraform.ResourceProvider {
+			return Provider()
+		},
+	})
+}
+```
+
+This establishes the main function to produce a valid, executable Go binary. The
+contents of the main function consume Terraform's `plugin` library. This library
+deals with all the communication between Terraform core and the plugin.
+
+Next, build the plugin using the Go toolchain:
+
+```shell
+$ go build -o terraform-provider-example
+```
+
+The output name (`-o`) is **very important**. Terraform searches for plugins in
+the format of:
+
+```text
+terraform-<TYPE>-<NAME>
+```
+
+In the case above, the plugin is of type "provider" and of name "example".
+
+To verify things are working correctly, execute the binary just created:
+
+```shell
+$ ./terraform-provider-example
+This binary is a plugin. These are not meant to be executed directly.
+Please execute the program that consumes these plugins, which will
+load any plugins automatically
+```
+
+This is the basic project structure and scaffolding for a Terraform plugin.
```

Comment:
Would be helpful to clarify the project structure and scaffolding with a `tree` or `ls` output.